File: app/modules/mail_inbox/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, field_validator

from app.modules.shared.schemas import DashboardModel

MailProvider = Literal["gmail", "outlook", "imap"]
MailSyncStatusValue = Literal["never_synced", "synced", "error", "disabled"]
MailFocusRuleKindValue = Literal["sender_email", "sender_domain", "keyword"]


class MailAccountCreateRequest(DashboardModel):
    provider: MailProvider
    address: str
    display_name: str | None = None
    imap_host: str | None = None
    imap_port: int | None = Field(default=None, ge=1, le=65535)
    imap_username: str | None = None
    credential: str | None = None

    @field_validator("address")
    @classmethod
    def _normalize_address(cls, value: str) -> str:
        normalized = value.strip().lower()
        if not normalized:
            raise ValueError("Address is required")
        return normalized


class MailAccountUpdateRequest(DashboardModel):
    enabled: bool | None = None
    display_name: str | None = None


class MailAccountResponse(DashboardModel):
    id: str
    provider: MailProvider
    address: str
    display_name: str | None
    enabled: bool
    sync_status: MailSyncStatusValue
    last_sync_at: datetime | None
    last_sync_error: str | None
    imap_host: str | None
    imap_port: int | None
    imap_username: str | None
    created_at: datetime
    updated_at: datetime


class MailAccountsResponse(DashboardModel):
    accounts: list[MailAccountResponse]


class MailAccountSyncResponse(DashboardModel):
    account: MailAccountResponse
    imported_count: int


class MailMessageIngestRequest(DashboardModel):
    account_id: str
    provider_message_id: str
    thread_id: str | None = None
    sender_email: str
    sender_name: str | None = None
    recipients: list[str] = Field(default_factory=list)
    subject: str
    snippet: str = ""
    received_at: datetime
    unread: bool = False
    starred: bool = False
    has_attachments: bool = False

    @field_validator("sender_email")
    @classmethod
    def _normalize_sender(cls, value: str) -> str:
        normalized = value.strip().lower()
        if not normalized:
            raise ValueError("Sender email is required")
        return normalized


class MailMessageResponse(DashboardModel):
    id: str
    account_id: str
    account_address: str
    account_provider: MailProvider
    provider_message_id: str
    thread_id: str | None
    sender_email: str
    sender_name: str | None
    recipients: list[str]
    subject: str
    snippet: str
    received_at: datetime
    unread: bool
    starred: bool
    has_attachments: bool
    focused: bool
    focus_label: str | None


class MailMessagesResponse(DashboardModel):
    messages: list[MailMessageResponse]


class MailFocusRuleCreateRequest(DashboardModel):
    kind: MailFocusRuleKindValue
    value: str
    label: str | None = None

    @field_validator("value")
    @classmethod
    def _normalize_value(cls, value: str) -> str:
        normalized = value.strip().lower()
        if not normalized:
            raise ValueError("Rule value is required")
        return normalized


class MailFocusRuleUpdateRequest(DashboardModel):
    enabled: bool


class MailFocusRuleResponse(DashboardModel):
    id: str
    kind: MailFocusRuleKindValue
    value: str
    label: str | None
    enabled: bool
    created_at: datetime
    updated_at: datetime


class MailFocusRulesResponse(DashboardModel):
    rules: list[MailFocusRuleResponse]

File: frontend/src/features/mail/schemas.ts
import { z } from "zod";

export const MailProviderSchema = z.enum(["gmail", "outlook", "imap"]);
export const MailFocusRuleKindSchema = z.enum(["sender_email", "sender_domain", "keyword"]);

export const MailAccountSchema = z.object({
  id: z.string(),
  provider: MailProviderSchema,
  address: z.string(),
  displayName: z.string().nullable(),
  enabled: z.boolean(),
  syncStatus: z.enum(["never_synced", "synced", "error", "disabled"]),
  lastSyncAt: z.string().nullable(),
  lastSyncError: z.string().nullable(),
  imapHost: z.string().nullable(),
  imapPort: z.number().nullable(),
  imapUsername: z.string().nullable(),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export const MailAccountsResponseSchema = z.object({
  accounts: z.array(MailAccountSchema),
});

export const MailAccountSyncResponseSchema = z.object({
  account: MailAccountSchema,
  importedCount: z.number(),
});

export const MailMessageSchema = z.object({
  id: z.string(),
  accountId: z.string(),
  accountAddress: z.string(),
  accountProvider: MailProviderSchema,
  providerMessageId: z.string(),
  threadId: z.string().nullable(),
  senderEmail: z.string(),
  senderName: z.string().nullable(),
  recipients: z.array(z.string()),
  subject: z.string(),
  snippet: z.string(),
  receivedAt: z.string(),
  unread: z.boolean(),
  starred: z.boolean(),
  hasAttachments: z.boolean(),
  focused: z.boolean(),
  focusLabel: z.string().nullable(),
});

export const MailMessagesResponseSchema = z.object({
  messages: z.array(MailMessageSchema),
});

export const MailFocusRuleSchema = z.object({
  id: z.string(),
  kind: MailFocusRuleKindSchema,
  value: z.string(),
  label: z.string().nullable(),
  enabled: z.boolean(),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export const MailFocusRulesResponseSchema = z.object({
  rules: z.array(MailFocusRuleSchema),
});

export type MailProvider = z.infer<typeof MailProviderSchema>;
export type MailFocusRuleKind = z.infer<typeof MailFocusRuleKindSchema>;
export type MailAccount = z.infer<typeof MailAccountSchema>;
export type MailMessage = z.infer<typeof MailMessageSchema>;
export type MailFocusRule = z.infer<typeof MailFocusRuleSchema>;
export type MailAccountsResponse = z.infer<typeof MailAccountsResponseSchema>;
export type MailAccountSyncResponse = z.infer<typeof MailAccountSyncResponseSchema>;
export type MailMessagesResponse = z.infer<typeof MailMessagesResponseSchema>;
export type MailFocusRulesResponse = z.infer<typeof MailFocusRulesResponseSchema>;

export type MailAccountCreatePayload = {
  provider: MailProvider;
  address: string;
  displayName?: string | null;
  imapHost?: string | null;
  imapPort?: number | null;
  imapUsername?: string | null;
  credential?: string | null;
};

export type MailFocusRuleCreatePayload = {
  kind: MailFocusRuleKind;
  value: string;
  label?: string | null;
};
